7 - 12: Add exercises and solutions (#31)
ONSdigital/r-rap-demo: exercises/02_modules.Rmd

head(df, 5)
```

## Step 1: Review the Monolithic Script

Below is a single script that identifies missing values and performs imputation. Your task is to refactor this into modular code.

**Tasks:**
- Convert the below into functions
- Extend the function to impute missing height and weight using group means (e.g., by gender or diagnosis)
- Add a new column to flag which values were imputed

```{r}
# Impute missing height and weight with column means
# Flag missing values
df$height_cm_imputed <- is.na(df$height_cm)
df$weight_kg_imputed <- is.na(df$weight_kg)

df <- group_by(df, gender)
df <- mutate(df,
             height_cm = ifelse(is.na(height_cm), mean(height_cm, na.rm = TRUE), height_cm),
             weight_kg = ifelse(is.na(weight_kg), mean(weight_kg, na.rm = TRUE), weight_kg)
             )
df <- ungroup(df)

df <- mutate(df,
             height_cm = ifelse(is.na(height_cm), mean(height_cm, na.rm = TRUE), height_cm),
             weight_kg = ifelse(is.na(weight_kg), mean(weight_kg, na.rm = TRUE), weight_kg)
             )
print(df)
```
Write your function(s) in the cell below:
```{r}

```
## Step 2: Add your functions to the pipeline

**Tasks:**

> - Add your functions to an appropriate module in `R/R_rap_demo`

Reviewer: ```suggestion
- Add your functions to an appropriate module in the `R/` folder
```

There are a few instances where 'R_rap_demo/R/' is in the instructions, so just need to go through and change those to `R/` like above

- Ensure your functions are used when running the pipeline

**Hints:**
- Add your functions into another function in the same module
- If you create a new module, import the module into the main.R script

## Step 3: Challenge & Reflection

**Tasks:**
- Compare the cleaned_data output with and without your changes

**Reflect:**
- How does modular code improve reproducibility and maintainability?
- What are the benefits of separating code into modules?
- How would you extend this pipeline to add new features or analyses?

**Bonus:**

> - Add unit tests for your new functions

Reviewer: As we have a module on unit tests, I think we could replace this with. "Try adding your own modules and functions to the pipeline".

Same with the python one as in the comment below

> - Document your modules with clear comments and docstrings

Reviewer: Can we have this in Tasks in Step 2? Its quite important so probably shouldn't come under bonus. Fully aware this is copied from the python one so can you add this to the list of changes in the python demo you have as well please

Also for R we should change docstrings to Roxygen documentation e.g.
```suggestion
- Document your modules with Roxygen documentation
```

## Step 4: Refactor Visualisation Code

> As an additional challenge, Add code to create visualisations. Place all visualisation functions in a separate module (e.g., `report.R`).

Reviewer: May come back to this one, but its another for the python list. Might be an idea to get users to create a separate module e.g. `vis.R` to add the visualisation functions to. For now though I think we'd just need to change the wording e.g.

```suggestion
As an additional challenge, Add code to create visualisations. Place all visualisation functions in the `report.R` module.
```
